### PR TITLE
Additional scheme endpoints for projects

### PR DESCRIPTION
**Pull request by Zyzyx**
See https://github.com/pycontribs/jira/issues/1294

### CODE REVIEW

File: jira/client.py [hunk 1]
```diff
@@ -2140,6 +2141,19 @@ def project_permissionscheme(self, project: str) -> PermissionScheme:
...
+    @translate_resource_args
+    def project_priorityscheme(self, project: str) -> PriorityScheme:
+        """Get a PriorityScheme Resource from the server.
+
+
+        Args:
+            project (str): ID or key of the project to get the PriorityScheme for
+
+        Returns:
+            PriorityScheme,: The priority scheme
```
Comment (adehad) [inline jira/client.py:2153]:
```suggestion
            PriorityScheme: The priority scheme
```

**Review by adehad — CHANGES_REQUESTED** [review]
Thanks @Zyzyx !

Don't suppose I can trouble you with a test case for this?
It would look very similar to: 
https://github.com/pycontribs/jira/blob/064a1cdf5f956a9da221ad88bbf1e356c0951b94/tests/resources/test_project.py#L204

File: jira/client.py (hunk 1 above)
Comment (Zyzyx) [reply jira/client.py:2153]: Sure. I'm struggling to get credentials for the CI Jira Cloud account though. I wasn't sure if I should create yet another pair of accounts (I see 2 pair already) or try to recover/reset credentials for the cloud CI. For local testing I'll use my own account.

Comment (adehad) [reply jira/client.py:2184]: Yep I can send those to you on element, but if your individual account works that is fine too.
It's a bit annoying at the moment as only branches created from this project run the Cloud CI (limitation of Github sharing the credentials it seems) but if the server tests pass that is good enough to start with.

File: jira/client.py [hunk 2]
```diff
@@ -3983,6 +4041,16 @@ def permissionschemes(self):
 
         return data["permissionSchemes"]
 
+    @lru_cache(maxsize=None)
+    def issuetypeschemes(self):
+
```
Comment (adehad) [inline jira/client.py:4046]:
```suggestion
    def issue_type_schemes(self) -> List[IssueTypeScheme]:
    """<docstring>"""
```

Just a reminder to use snake case for function names, to add the right typehints and to add a sensible docstring.

Comment (adehad) [reply jira/client.py]: Reminder: Docstrings should conform to the Google docstring style and include return type, I can still see some rogue commas in the return sections too

File: examples/maintenance.py [hunk 3]
```diff
@@ -14,13 +14,13 @@
 logging.getLogger("urllib3").setLevel(logging.INFO)
 logging.getLogger("jira").setLevel(logging.DEBUG)
 
-CI_JIRA_URL = os.environ["CI_JIRA_URL"]
+CI_JIRA_URL = os.environ["CI_JIRA_URL"]  # https://pycontribs.atlassian.net
```
Comment (adehad) [inline examples/maintenance.py:17]:
```suggestion
CI_JIRA_URL = os.environ["CI_JIRA_URL"]
```

File: jira/client.py [hunk 4]
```diff
@@ -3984,7 +4043,18 @@ def permissionschemes(self):
         return data["permissionSchemes"]
 
     @lru_cache(maxsize=None)
-    def issuesecurityschemes(self):
```
Comment (adehad) [inline jira/client.py:3987]: Unfortunately, as this is an API that is widely used, we cannot simply change these public methods.
If you are keen to make this change, my suggestion would be to: make a separate pull request where you deprecate this existing function name, and make it call the new function name

File: jira/client.py [hunk 5]
```diff
@@ -4014,7 +4084,7 @@ def avatars(self, entity="project"):
         return data["system"]
 
     @lru_cache(maxsize=None)
-    def notificationschemes(self):
+    def notification_schemes(self):
```
Comment (adehad) [inline jira/client.py:4087]: as above regarding deprecation

File: jira/client.py [hunk 6]
```diff
@@ -4032,7 +4102,7 @@ def screens(self):
         return data["values"]
 
     @lru_cache(maxsize=None)
-    def workflowscheme(self):
+    def workflow_scheme(self):
```
Comment (adehad) [inline jira/client.py:4105]: as above regarding deprecation

File: tests/resources/test_project.py [hunk 7]
```diff
@@ -201,6 +202,29 @@ def test_project_roles(self):
...
+    def test_project_priority_scheme(self):
+        priorityscheme = self.jira.project_priority_scheme(self.project_b)
+        self.assertEqual(priorityscheme.name, "Default priority scheme")
+
+    def test_project_notification_scheme(self):
+        notificationscheme = self.jira.project_notification_scheme(self.project_b)
+        self.assertEqual(notificationscheme.name, "Default Notification Scheme")
+
+    def test_project_issue_security_level_scheme(self):
+        # 404s are thrown when a project does not have an issue security scheme
```
Comment (adehad) [inline tests/resources/test_project.py:218]: excellent comment

File: tests/resources/test_issue_type_scheme_associations.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
+from tests.conftest import JiraTestCase
+
+
+class IssueTypeSchemeAssociationTests(JiraTestCase):
+    def test_scheme_associations(self):
+        allSchemes = self.jira.issue_type_schemes()
```
Comment (adehad) [inline tests/resources/test_issue_type_scheme_associations.py:6]:
```suggestion
        all_schemes = self.jira.issue_type_schemes()
```

is the preferred way for variables, older code isn't as consistent, but let's try and be consistent with the newly added stuff

File: tests/resources/test_issue_type_scheme_associations.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+        # there should be more than 1 scheme
+        self.assertGreaterEqual(len(allSchemes), 2)
+        g2g = False
```
Comment (adehad) [inline tests/resources/test_issue_type_scheme_associations.py:9]:
```suggestion
        is_a_scheme_associated_with_a_project = False
```
I would avoid acronyms or abbreviations like this in variable names, consider more descriptive names that also better confer the meaning of your test.

File: tests/resources/test_issue_type_scheme_associations.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+        for scheme in allSchemes:
+            associations = self.jira.get_issue_type_scheme_associations(scheme["id"])
+            # As long as one of these schemes is associated with a project-like object
+            # we're probably ok.
+            if len(associations) > 0:
+                print(associations[0].get("id"))
```
Comment (adehad) [inline tests/resources/test_issue_type_scheme_associations.py:15]: no point printing this

**Review by adehad — CHANGES_REQUESTED** [review]
Awesome stuff, a few minor points, but looking good

File: tox.ini [hunk 11]
```diff
@@ -18,6 +19,7 @@ python =
     3.7: py37
     3.8: py38
     3.9: py39
+    3.10: py310
```
Comment (adehad) [inline tox.ini:22]: I tried this on my fork, and it didn't work... did you manage to get this working locally?

https://github.com/adehad/jira/actions/runs/2044892461

Comment (Zyzyx) [reply tox.ini]: I'll back this out for now. Maybe whether a separate PR.

File: jira/client.py [hunk 12]
```diff
@@ -3984,7 +4043,18 @@ def permissionschemes(self):
...
+
+        url = self._get_url("issuetypescheme")
+
+        r = self._session.get(url)
+        data: Dict[str, Any] = json_loads(r)
+
+        return data["schemes"]
+
+    @lru_cache(maxsize=None)
+    def issue_securityschemes(self):
```
Comment (adehad) [inline jira/client.py:4057]:
```suggestion
    def issuesecurityschemes(self):
```

is the original name (unfortunately)

**Review by adehad — CHANGES_REQUESTED** [review]
Almost there!

File: jira/client.py (hunk 12 above)
Comment (Zyzyx) [reply jira/client.py:4057]: whoops, good catch

File: jira/client.py [hunk 13]
```diff
@@ -4069,6 +4139,12 @@ def delete_permissionscheme(self, id: str):
         self.permissionschemes.cache_clear()
         return data
 
+    def get_issue_type_scheme_associations(self, id):
```
Comment (adehad) [inline jira/client.py:4142]: As this is a public facing function this needs typehints and docstring.

This looks like it should return a `List[Project]`

https://docs.atlassian.com/software/jira/docs/api/REST/8.13.6/#api/2/issuetypescheme-getAssociatedProjects

```suggestion
    def get_issue_type_scheme_associations(self, id: str) -> List[Project]:
    """For the specified issue type scheme, returns all of the associated projects. (Admin required)
    
    Args:
        id (str): The issue type scheme id.
        
    Returns:
        List[Project]: Associated Projects for the Issue Type Scheme.
    """
```

File: jira/client.py [hunk 14]
```diff
@@ -3983,6 +4042,17 @@ def permissionschemes(self):
 
         return data["permissionSchemes"]
 
+    @lru_cache(maxsize=None)
+    def issue_type_schemes(self) -> List[IssueTypeScheme]:
+        """get all issue type schemes defined"""
```
Comment (adehad) [inline jira/client.py:4047]: This needs the docstring in the Google style

```suggestion
        """Get all issue type schemes defined (Admin required).
        
        Returns:
            List[IssueTypeScheme]: All the Issue Type Schemes available to the currently logged in user.
        """
```

File: jira/client.py [hunk 15]
```diff
@@ -2140,6 +2173,32 @@ def project_permissionscheme(self, project: str) -> PermissionScheme:
         """
         return self._find_for_resource(PermissionScheme, project)
 
+    @translate_resource_args
+    def project_priority_scheme(self, project: str) -> PriorityScheme:
+        """Get a PriorityScheme Resource from the server.
+
```
Comment (adehad) [inline jira/client.py:2179]:
```suggestion
```

redundant new line

File: jira/client.py [hunk 16]
```diff
@@ -2140,6 +2173,32 @@ def project_permissionscheme(self, project: str) -> PermissionScheme:
...
+        Returns:
+            PriorityScheme: The priority scheme
+        """
+        return self._find_for_resource(PriorityScheme, project)
+
+    @translate_resource_args
+    def project_workflow_scheme(self, project: str) -> WorkflowScheme:
+        """Get a WorkflowScheme Resource from the server.
+
+
```
Comment (adehad) [inline jira/client.py:2193]:
```suggestion
```

redundant new line

File: jira/client.py [hunk 17]
```diff
@@ -2127,6 +2132,34 @@ def votes(self, issue: str) -> Votes:
...
+        Returns:
+            IssueSecurityLevelScheme: The issue security level scheme
+        """
+        return self._find_for_resource(IssueSecurityLevelScheme, project)
+
+    @translate_resource_args
+    def project_notification_scheme(self, project: str) -> NotificationScheme:
+        """Get a NotificationScheme Resource from the server.
+
```
Comment (adehad) [inline jira/client.py:2153]:
```suggestion
```

redundant new line

File: jira/client.py [hunk 18]
```diff
@@ -2127,6 +2132,34 @@ def votes(self, issue: str) -> Votes:
         """
         return self._find_for_resource(Votes, issue)
 
+    @translate_resource_args
+    def project_issue_security_level_scheme(
+        self, project: str
+    ) -> IssueSecurityLevelScheme:
+        """Get a IssueSecurityLevelScheme Resource from the server.
+
+
```
Comment (adehad) [inline jira/client.py:2141]:
```suggestion
```

redundant new line

**Review by adehad — CHANGES_REQUESTED** [review]
Sorry just being a bit more pedantic here for the public facing functions.

Some other minor superificial changes too

**Review by adehad — APPROVED** [review]
